tests/core/io/test_tcp_server.h

```diff
@@ -0,0 +1,210 @@
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#ifndef TEST_TCP_SERVER_H
+#define TEST_TCP_SERVER_H
+
+#include "core/io/stream_peer_tcp.h"
+#include "core/io/tcp_server.h"
+#include "tests/test_macros.h"
+
+namespace TestTCPServer {
+
+const int port = 12345;
+const IPAddress localhost("127.0.0.1");
+const int sleep_duration = 100000;
+
+Ref<TCPServer> create_server(const IPAddress &p_address, int p_port) {
+	Ref<TCPServer> server;
+	server.instantiate();
+
+	REQUIRE_EQ(server->listen(port, localhost), Error::OK);
+	REQUIRE(server->is_listening());
+	CHECK_FALSE(server->is_connection_available());
+
+	return server;
+}
+
+Ref<StreamPeerTCP> create_client(const IPAddress &p_address, int p_port) {
+	Ref<StreamPeerTCP> client;
+	client.instantiate();
+
+	REQUIRE_EQ(client->connect_to_host(localhost, port), Error::OK);
+	CHECK_EQ(client->get_connected_host(), localhost);
+	CHECK_EQ(client->get_connected_port(), port);
+	CHECK_EQ(client->get_status(), StreamPeerTCP::STATUS_CONNECTING);
+
+	return client;
+}
+
+Ref<StreamPeerTCP> accept_connection(Ref<TCPServer> &server) {
+	REQUIRE(server->is_connection_available());
+	Ref<StreamPeerTCP> client_from_server = server->take_connection();
+	REQUIRE_MESSAGE(client_from_server.is_valid(), "A Stream Peer TCP from the TCP Server should not be a null pointer.");
+	CHECK_EQ(client_from_server->get_connected_host(), localhost);
+	CHECK_EQ(client_from_server->get_status(), StreamPeerTCP::STATUS_CONNECTED);
+
+	return client_from_server;
+}
+
+TEST_CASE("[TCPServer] Instantiation") {
+	Ref<TCPServer> server;
+	server.instantiate();
+
+	REQUIRE_MESSAGE(server.is_valid(), "A TCP Server created should not be a null pointer.");
+	CHECK_EQ(false, server->is_listening());
+}
+
+TEST_CASE("[TCPServer] Accept a connection and receive/send data") {
+	Ref<TCPServer> server = create_server(localhost, port);
+	Ref<StreamPeerTCP> client = create_client(localhost, port);
+
+	// Required to get the connection properly established.
+	OS::get_singleton()->delay_usec(sleep_duration);
+
+	Ref<StreamPeerTCP> client_from_server = accept_connection(server);
```

Comment:
As an example for https://github.com/godotengine/godot/pull/97262#discussion_r1805889374 here you probably want to do something like (in pseudocode):

```pseudocode
time = get_ticks()
while (!is_connection_available()) {
    poll()
    sleep(some_ms)
    if (get_ticks() - time < max_wait_ms) {
        break; // AND FAIL
    }
}
```